### ISSUE TITLE
FBXLoader.js should use the "min" function and not the "max" function when limiting the penumbra value.

### ISSUE THREAD

**Issue by roschler** [mrdoob/three.js#24910]
**Describe the bug**

`FBXLoader.js` should be use the "min" function instead of the "max" function for the _penumbra_ limiting statement.

***Code***

Here is the suspect code.  The code uses the "max" function when it should use "min" to limit the _penumbra_ to a value less than or equal 1. 

https://github.com/mrdoob/three.js/blob/f54eb5b0bf7baa7132acc26c799bbb16bda136a8/examples/jsm/loaders/FBXLoader.js#L1206-L1217

Also, wouldn't the first statement better as follows?:
```js
        penumbra = THREE.MathUtils.degToRad( lightAttribute.OuterAngle.value  / 360.0 );
	penumbra = Math.min( penumbra, 1 );
```
By dividing the FBX model angle value by 360 you rescale it to 0, 1.

**Comment by Mugen87** [comment]
Penumbra in `three.js` is defined as follows:

>  Percent of the spotlight cone that is attenuated due to penumbra. Takes values between zero and 1. Default is zero.

I don't think this is right to use `Math.min()` instead of `Math.max()` since you would end up with values outside of the `[0,1]` range.

To me, the code in the loader looks correct. It's probably best if you share an FBX assets that demonstrates the issue.

**Comment by roschler** [comment]
> To me, the code in the loader looks correct. It's probably best if you share an FBX assets that demonstrates the issue.

Consider the current code in `FBXLoader.js`:
```js
 	penumbra = Math.max( penumbra, 1 ); 
``` 	
If `penumbra` is greater than 1 then `Math.max()`, which returns the maximum value between a pair of provided input values, will return a value greater than 1.  As you said, the value has to be less than or equal to one (and greater than or equal to zero), so that results in an invalid value if `penumbra` is greater than 1.  If you use use the `min()` function instead, and the value calculated for `penumbra` is greater than 1, The `min()` function will return 1.

From the MDN docs for `Math.min()`:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/max

```
The Math.max() function returns the largest of the numbers given as input parameters, or -[Infinity](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Infinity) if there are no parameters.
```

**Comment by WestLangley** [comment]
Guessing at something that may be reasonable...   /ping @looeee 

Set the FBX `innerAngle` and `outerAngle` to 45 degrees if either are undefined. (Confirm these are correct default values.)

Make sure 0 < `innerAngle` < `outerAngle` < 90 deg.

Then,
```js
spotlight.angle = MathUtils.degToRad( outerAngle );

spotlight.penumbra = innerAngle / outerAngle;
```

@Mugen87 Does that seem reasonable to you?

@roschler Would you like to file a PR?

**Comment by roschler** [comment]
> @roschler Would you like to file a PR?

You might want to look at what I would suggest first.  I did some digging on the FBX file format in regards to the `InnerAngle` and `OuterAngle` properties related to a spotlight defined in the file.  I believe now that the existing `FBXLoader.js` code has those two properties being used in an inverted manner compared to their intended use.  In other words, the two field names should be swapped with each other.  For example the current `FBXLoader` code assigns the `InnerAngle` to the `THREE.SpotLight` cone angle, when the `OuterAngle` should be used instead.  

I say this because of what I found in these FBX documents:

---

FBX Docs  URLs

If you read the first document you will see the legacy field names for `InnerAngle` and `OuterAngle`, which used to be `HotSpot` and `ConeAngle`.  This document defines ConeAngle as the outer angle, which is why it was renamed to OuterAngle later.  It defines HotSpot as the inner angle, which is why it was renamed to InnerAngle later:

https://download.autodesk.com/us/fbx/sdkdocs/fbx_sdk_help/files/fbxsdkref/class_k_fbx_light.html

This doc confirms the meaning of the InnerAngle and OuterAngle properties:

http://docs.autodesk.com/FBX/2014/ENU/FBX-SDK-Documentation/index.html?url=files/GUID-8EA6A733-8B78-4698-85A7-F19F426E9DD1.htm,topicNumber=d30e2057

Therefore the `OuterAngle` value should be assigned to the `THREE.SpotLight` _angle_ parameter, and then you have to derive the `penumbra` value from the ratio of the FBX `InnerAngle` and `OuterAngle` field values.

Because of this, I would restructure the code as shown below.  You tell me if you still want me to submit a PR.  Note, before I did that, I would require an FBX that has a spotlight defined in it to test this code, since this could is _not_ tested:

---

```js
// The default outer angle for the spotlight is 45 degrees,
//  expressed in radians because that is the format that
//  ThreeUS uses.
let angle = Math.PI / 4;

// Does the FBX file define an OuterAngle?
if ( lightAttribute.OuterAngle !== undefined ) {

    // Yes, the FBX file has a defined outer angle value,
    // so we should use it as the light cone angle.  
    // Limit the value to a value between 0 and 180 degrees.
    const angleDegrees = Math.max( 0, Math.min( 180, lightAttribute.OuterAngle.value ) );

    // Convert the value to radians because that is
    // the format ThreeJS uses.
    angle = THREE.MathUtils.degToRad( angleDegrees );
}

// ThreeJS expresses the penumbra as a percentage
//  of the angle value for the spotlight.  We set
//  the default value for the penumbra to 50%.  With
//  this value, as you move away from the light
//  cone's center , the light starts decaying at the
//  halfway point of the cone's radius.
let penumbra = 0.5;

// Does the FBX file define an InnerAngle value and
//  is that value less than the defined OuterAngle?
if ( lightAttribute.InnerAngle !== undefined
        && lightAttribute.OuterAngle !== undefined // A declared inner angle is only valid if the FBX file declared an outer angle.
        && lightAttribute.OuterAngle.value >= 0 // A negative inner angle is not valid.
        && lightAttribute.OuterAngle.value > 0 // This is to avoid a divide by zero error below.
        && lightAttribute.InnerAngle.value <= lightAttribute.OuterAngle.value // An inner angle that is greater than the outer angle is invalid.
    ) {

    // Yes, the FBX file has a defined inner angle value.
    //  Determine the penumbra percentage by finding
    //  the ratio between the inner and outer angle
    //  values defined in the FBX file, and subtracting
    //  that value from 1 to find the percentage of the
    //  light cone that should be part of the penumbra.
    let rawPenumbra = 1 - ( lightAttribute.InnerAngle.value / lightAttribute.OuterAngle.value );

    // This should not be necessary given the logic constraints
    //  encoded in the statements above, but we explicitly limit
    //  the value to the mandated range of 0 and 1 for
    //  the penumbra just in case.
    penumbra = Math.max(0, Math.min(1, rawPenumbra))
}
```

**Comment by looeee** [comment]
It's a while since I looked at this stuff so here's the relevant docs that I could find:

For reference here's [the FBX docs for inner/outer angle](https://help.autodesk.com/view/FBX/2017/ENU/?guid=__cpp_ref_class_fbx_light_html) (can't link directly, search on the page for innerangle):

> InnerAngle: This property handles the light inner cone angle (in degrees). Also know as the HotSpot. Default value is 45.0
> OuterAngle: This property handles the light outer cone angle (in degrees). Also known as the Falloff. Default value is 45.0

The [three.js concepts of angle and penumbra](https://threejs.org/docs/?q=spotli#api/en/lights/SpotLight):

> angle - Maximum angle of light dispersion from its direction whose upper bound is Math.PI/2.
> penumbra - Percent of the spotlight cone that is attenuated due to penumbra. Takes values between zero and 1. Default is zero.

Also, here's the description of [falloff and hotspot from the 3ds max docs](https://knowledge.autodesk.com/support/3ds-max/learn-explore/caas/CloudHelp/cloudhelp/2021/ENU/3DSMax-Lighting-Shading/files/GUID-4A96BEC8-66E5-4DA4-BD78-5E2872A3D2AD-htm.html):

> Hotspot/Beam/_innerangle_: Adjusts the angle of a light's cone. The Hotspot value is measured in degrees. Default=43.0.
> Falloff/Field/_outerangle_: Adjusts the angle of a light's falloff. The Falloff value is measured in degrees. Default=45.0.
> For photometric lights, the Field angle is comparable to the Falloff angle. It is the angle at which the light's intensity has fallen to zero.

From [brownbagfilms.com](https://www.brownbagfilms.com/labs/entry/basic-lighting-in-3ds-max#:~:text=The%20spotlight%20also,disc%20or%20rectangle.) talking about the 3ds max spotlight:

> The spotlight also has its own parameters section. The main things to take into account here are the hotspot/beam and falloff settings. These are fairly straightforward, just think of it as two cones, the inner one being the full brightness (Hotspot) of the spotlight and the outer being the penumbra or falloff, the outer edge of which the light falls off to zero. 

So it seems ok to me that `outerangle = spotlight.angle` and `innerangle = spotlight.penumbra`, with appropriate conversions to radians and ratio. @Mugen87 and @WestLangley, is it correct that the three.js penumbra sets an inner cone of full brightness? 

@roschler given that I think your code is mostly ok, except for the penumbra default. In FBX, it defaults to 45 degrees. That would give a penumbra of 1 in three.js _if_ the angle is also 45 degrees. If it's not, we need to scale it to get the eqivalant ratio for other angles.

**Comment by WestLangley** [comment]
>is it correct that the three.js penumbra sets an inner cone of full brightness?

Yes, but the three.js penumbra attenuation formula probably should be revisited.
